src/api/app/models/project.rb

```diff
@@ -1643,4 +1697,125 @@ def to_param
     name
   end
 
+  # the following code checks if the target project of a linked project exists or is not readable by user
+  def self.check_target_of_link(project_name, request_data)
+    request_data.elements('link') do |e|
+      # permissions check
+      target_project_name = e.value('project')
+      target_project = Project.get_by_name(target_project_name)
+
+      # The read access protection for own and linked project must be the same.
+      # ignore this for remote targets
+      if target_project.class == Project &&
+          target_project.disabled_for?('access', nil, nil) &&
+          !FlagHelper.xml_disabled_for?(request_data, 'access')
+        raise ProjectReadAccessFailure, 'project links work only when both projects have same read access ' +
+            "protection level: #{project_name} -> #{target_project_name}"
+      end
+      logger.debug "project #{project_name} link checked against #{target_project_name} projects permission"
+    end
+  end
+
+  def self.update_project_meta_maintains(request_data)
+    # Check if used maintains statements point only to projects with write access
+    request_data.elements('maintenance') do |m|
+      m.elements('maintains') do |r|
+        target_project_name = r.value('project')
+        target_project = Project.get_by_name(target_project_name)
+        unless target_project.class == Project && User.current.can_modify_project?(target_project)
+          raise ModifyProjectNoPermission, "No write access to maintained project #{target_project_name}"
+        end
+      end
+    end
+  end
+
+
+  def check_and_remove_repositories(params, remove_repositories)
+    # find linking repos which get deleted
+    linkingRepositories = []
+    linkingTargetRepositories = []
+
+    remove_repositories.each do |repo|
+      linkingRepositories += repo.linking_repositories
+      linkingTargetRepositories += repo.linking_target_repositories
+    end
+
+    unless params[:force] && !params[:force].empty?
+      check_dependent_repositories!(linkingRepositories,
+                                    'Unable to delete repository; following repositories depend on this project:')
+
+      check_dependent_repositories!(linkingTargetRepositories,
+                                    'Unable to delete repository; following target repositories depend on this project:')
```

Comment:
Does this make sense? linkingRepositories and linkingTargetRepositories should contain the same content, right?